File: python/linkedlist/reorderList.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @return a ListNode
    def reorderList(self, head):
        if head is None or head.next is None:
            return head

        fast = head
        slow = head

        while fast.next and fast.next.next is not None:
            fast = fast.next.next
            slow = slow.next

        second_half = slow.next
        slow.next = None
        first_half = head 

        dummy = ListNode(0)
        dummy.next = second_half
        start = dummy
        first = start.next 
        while first.next:
            t0 = start.next
            t1 = first.next.next
            start.next = first.next
            first.next.next = t0
            first.next = t1

        second_half = dummy.next

        result = ListNode(0)
        result.next = first_half
        while second_half:
            tfh = first_half.next
            tsh = second_half.next
            first_half.next = second_half
            second_half.next = tfh
            first_half = tfh
            second_half = tsh

        return result.next


if __name__ == "__main__":
    solution = Solution()
    l1 = ListNode(1)
    l2 = ListNode(2)
    l3 = ListNode(3)
    l4 = ListNode(4)
    l5 = ListNode(5)
    l1.next = l2
    l2.next = l3
    l3.next = l4
    l4.next = l5

    p = solution.reorderList(l1)
    while p is not None:
        print p.val
        p = p.next

File: java/ReorderList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public void reorderList(ListNode head) {
        if (head == null || head.next == null) {
            return;
        }
        
        ListNode slow = head;
        ListNode fast = head;
        while (fast.next != null && fast.next.next != null) {
            fast = fast.next.next;
            slow = slow.next;
        }
        
        ListNode h2 = slow.next;
        slow.next = null;
        
        ListNode h1 = head;
        h2 = this.reverse(h2);
        while (h1 != null && h2 != null) {
            ListNode tmp1 = h1.next;
            ListNode tmp2 = h2.next;
            h1.next = h2;
            h2.next = tmp1;
            h1 = tmp1;
            h2 = tmp2;
        }
    }
    
    public ListNode reverse(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode tail = dummy.next;
        while (tail.next != null) {
            ListNode temp = tail.next;
            tail.next = temp.next;
            temp.next = dummy.next;
            dummy.next = temp;
        }
        
        return dummy.next;
    }
}
